### DISCUSSION TITLE
Some experiments with UI and dom interactions

### DISCUSSION THREAD

**Discussion by Neon22** [pyscript/pyscript#1586] [Proposals]
Some thoughts on dom interactions - Element(), display(), @when, `py-event`

In my experimentation with pyscript and manipulating the DOM I have some questions about best practices.
I've been trying to do something a little complex here: https://d5127013-0d63-43cd-9926-95380d21cf67.pyscriptapps.com/dab7c27f-882a-4e8d-b88d-adf7ef49fa35/latest/
involving svg over img and dynamic reactive updates of user interaction for color harmony model selection.

1. Needs
 For a reactive style of UI these are the kinds of actions I need to do:
   - mechanism to build an initial UI
     - existing mechanism of redefining the innerHtml works well.
   - mechanism for python code to be called when events occur on dom elements 
     - @when and py-event hooks work with some issues(below)
   - mechanism to change the dom dynamically. E.g.
     - add fields because an array has changed in size
     - toggle visibility of parts of the dom
     - change the events, or event params, on an element
   - in this style of reactivity, its enough to change classes which are predefined in a css file rather than directly create css properties
   - in the future a single file component (grouping html,css,python) would help to make all this more modular.
   - similarly a templating model for the html
     - IMHO Vue3 and the composition API is a good reference for design/useability. Whereas `React` appears almost as a new language.
 
**Existing mechanism - strengths and weaknesses**
 
1. Element()
   - Element() works well for html ids but to do additional work you need to use Element().element
      - E.g. these are equivalent js and pyscript mechanisms:
	```
	block_div = js.document.getElementById("blocks") # where blocks is an id
    bars = block_div.querySelectorAll(".color-bar")
	```
	```
	block_div = Element("blocks") # where blocks is an id
    bars = block_div.element.querySelectorAll(".color-bar")
	```
      - It could be argued that the Element should know that the querySelectorAll method needs to be applied to the internal .element and so not need a .element method?
      - Equally there is the argument that its a bad idea to hide what's really going on.
	 - The .select() method only does querySelector() so maybe its not neccessary to have this method (or add all the methods) when you can use any of the js methods by operating on .element?
  
   - `Element()` will succeed even when the id does not exist. so in these cases checking for existence of .element is required to verify.
   - Element only works with ids and not classes.
   - An altenative approach might be to use something like `Select()` which returns an Element (or list of Elements).
      - Element() would then also be used as a constructor for making new elements prior to adding them into the dom ?
      - This might be a bridge to Single File Components where a dictionary(?) could hold the html,css,python code in a single file
      - Element.clone would then become more useful when constructing the UI programmatically (rather than through innerHtml replacement).
      - E.g. I have an svg and I want to add a unique id each time I add a new copy of it into the document,
	    so I can find it on user-selection and for multiple updates. Currently done as a brute string insertion into the svg block.
 
2. When inserting into the dom:
    - `blocks_div.element.innerHTML = "<div></div>"` works very well. (but maybe the same issue with .element above?)
	- The Element.write method seems the same as `display()` (but maybe display is for quick single use insertions).
	- However these two are not equivalent:
	  - `blocks_div.element.innerHTML = inner_html`
        `blocks_div.write(inner_html)`
		where inner_html is a string of html stuff e.g. `<div></div>`
	- BTW the `_repr_svg_` style of encapsulating mime types is very nice and encourages the benefits of classing these elements in python code.

3. display(*values, target="idname") Unfortunately display can't take an Element that has already been found like this:
      ```python
         block_div = Element("blocks")
         display(*values, target=block_div)
      ```
	 therefore more specific selection methods can't be used.

4. BTW this works, which is great: `areas = block_div.element.querySelectorAll('[class*="area"]')`
	It returns a list of e.g.`[object SVGPathElement]` and only looks under the block_div for the classes with 'area' in the classname.
	Clearly working for SVG as well as HTML elements.
	This means the GPS style of doing CSS (IMHO arguably more programming oriented than the BEM model) is easy to do.

5. Insertions and removals in the dynamic dom.
I've been changing the UI after initial creation using the following successfully:
```python
	block_frag = '<div blahblah"> <div blah></div> </div>'
	harmony = Element("harmony")
	# for HTML
	color_blocks = harmony.element.querySelectorAll(".color-block")
	# - returns list of [object HTMLDivElement]
	# to remove
	color_blocks[2].remove()
	# and to add:
	color_blocks[i].insertAdjacentHTML("afterend", block_frag)
    
	# for SVG
	svg = harmony.element.querySelector("#%s-ball-svg"%(id))
	# - returns an [object SVGSVGElement]
	# to remove
	svg.removeChild(svg.childNodes[-1]) # to pop off last svg element
	# to add is more complex and could do with a helper maybe
	newcircle = js.document.createElementNS('http://www.w3.org/2000/svg', 'circle');
	hue_circle_attr = [("cx","74"),("cy","38"),("r","6"),("transform","rotate(-90, 42, 42)"),("fill","none")]
	for attr,val in hue_circle_attr:
		newcircle.setAttribute(attr,val)
	svg.appendChild(newcircle)
```

6. py-events and @when
 - `@when()` seems to be doing a similar job to js.addEventHandler() but has two issues for me:
   - it fires on any child of the selector which means extra code is required to differentiate between parent and child with different classes, and to then act on the correct one.
   - it works only on page load. So additions to the dom do not get event messages. This is a problem if using this mechanism to handle user events and I add another element to my array.
   - This means I can't use @when on any element I intend to change in the dom.
   
 - py-events work well but, once inserted, are hard to change the parameters of. E.g.
   - I can have a `py-click=foo(var1)` where var is calculated on intial insertion of the innerHtml.
   - But its very hard to do any calculations when updating the argument obtained by target.getAttribute('py-click') because there is no notion of `this` when referring to the specific element I am updating.
     - This means I have to add `Element()` selection and a complex selector based on my knowledge of the exact location of this element(was class only) in the dom, or I have to add an id to every element I wish to manipulate this way.
     - Its also generating garbage to create and discard the `Element()` everytime this interaction occurs. 
   - so to get around this for now, I have to add an id to all such elements and use a `global Element()` to avoid excessive garbage creation.
   - so possibly a different mechanism is required, or its an argument for a global `this` to deal with events.

 - another possible garbage issue: Pyiodide has a `pyodide.ffi.wrappers.add_event_listener(elt, event, listener)` wrapper for JavaScript’s addEventListener() which automatically manages the lifetime of a JsProxy corresponding to the listener parameter.
   - but what happens if I use micropython in the near future. Will I be leaving a lot of garbage events still attached even though the dom elements may have been removed ?
   - IWBNI micropython based UI manipulation worked as well as Pyiodide.

I am not certain of what the 'right thing' to do is for any of the above.

**Comment by fpliger** [comment]
Hi @Neon22 , thanks for this. Great write-up and very useful to get use cases and opinions from our users! (Also, sorry for the slow reply... been busy on several fronts)

The good news is that I've been wanting to improve the higher level Web APIs we offer for some time and been also prototyping a bit. So this is great! 


**Some context first:** Tbh, I don't like the current `Element` API . I've pulled it in an afternoon without much thinking and since it was good enough for a few things, it just remained there because we (the maintainers of the project) were busy on other things with higher priority.

** The vision around the topic **: Ideally, I'd like us to provide a good python foundation API that allows others to build more abstract APIs on top of it. In that sense, I think that PyScript should come with a small PyScript `stdlib` that adds to the Python `stdlib` and we should be careful in adding things to that. As the community builds new libraries/framework on top of PyScript [and/or the new API] we can evaluate if it makes sense to actually pull into [or remove out of] that `stdlib` based on our community feedback, popularity, etc... Ideally, that `stdlib` only really loads things if our users are using them (in the web of things we need to be way more careful into not being too generous around loading things and lean more towards "less is more"

There's a lot of goodness to unpack here so bear with me :)

> 1. Needs
>  For a reactive style of UI these are the kinds of actions I need to do:
>    - mechanism to build an initial UI

Can you elaborate on this? You mean custom "pre-build" Elements that add already styled elements to the page? A template system? An example with pseudo-code could be helpful actually

>      - existing mechanism of redefining the innerHtml works well.
>    - mechanism for python code to be called when events occur on dom elements 
>      - @when and py-event hooks work with some issues(below)
>    - mechanism to change the dom dynamically. E.g.
>      - add fields because an array has changed in size
>      - toggle visibility of parts of the dom
>      - change the events, or event params, on an element

All of that makes a sense 👍 

>    - in this style of reactivity, its enough to change classes which are predefined in a css file rather than directly create css properties

RIght. Ultimately, we should probably support both things in a pythonic way

>    - in the future a single file component (grouping html,css,python) would help to make all this more modular.
>    - similarly a templating model for the html
>      - IMHO Vue3 and the composition API is a good reference for design/useability. Whereas `React` appears almost as a new language.

Can you provide pseudo-code examples? I think I get what you are saying but I want to make sure. 

On this "higher-level templating model", I think this probably belongs to frameworks on top of PyScript rather than on PyScript itself first, but maybe I'm not fully getting what you have in mind.

> 
> **Existing mechanism - strengths and weaknesses**
>  
> 1. Element()
>    - Element() works well for html ids but to do additional work you need to use Element().element
>       - E.g. these are equivalent js and pyscript mechanisms:
> 	```
> 	block_div = js.document.getElementById("blocks") # where blocks is an id
>     bars = block_div.querySelectorAll(".color-bar")
> 	```
> 	```
> 	block_div = Element("blocks") # where blocks is an id
>     bars = block_div.element.querySelectorAll(".color-bar")
> 	```
>       - It could be argued that the Element should know that the querySelectorAll method needs to be applied to the internal .element and so not need a .element method?
>       - Equally there is the argument that its a bad idea to hide what's really going on.
> 	 - The .select() method only does querySelector() so maybe its not neccessary to have this method (or add all the methods) when you can use any of the js methods by operating on .element?
>   
>    - `Element()` will succeed even when the id does not exist. so in these cases checking for existence of .element is required to verify.
>    - Element only works with ids and not classes.
>    - An altenative approach might be to use something like `Select()` which returns an Element (or list of Elements).
>       - Element() would then also be used as a constructor for making new elements prior to adding them into the dom ?
>       - This might be a bridge to Single File Components where a dictionary(?) could hold the html,css,python code in a single file
>       - Element.clone would then become more useful when constructing the UI programmatically (rather than through innerHtml replacement).
>       - E.g. I have an svg and I want to add a unique id each time I add a new copy of it into the document,
> 	    so I can find it on user-selection and for multiple updates. Currently done as a brute string insertion into the svg block.
>  

Agree with most (if not all) of the above. In fact, I've started experimenting with a new API that should account for all of this (I'll double-check and see if I'm missing something). It's very early on but you can check it out [here](https://pyscript.com/@7589ccb7-f838-4034-abea-5550f0e8d949/pyweb/latest) here in the meantime... (keep in mind that it's still very early on and still in the discovery/exploration/brainstorming phase) but early feedback is also very good.

> 2. When inserting into the dom:
>     - `blocks_div.element.innerHTML = "<div></div>"` works very well. (but maybe the same issue with .element above?)
> 	- The Element.write method seems the same as `display()` (but maybe display is for quick single use insertions).
> 	- However these two are not equivalent:
> 	  - `blocks_div.element.innerHTML = inner_html`
>         `blocks_div.write(inner_html)`
> 		where inner_html is a string of html stuff e.g. `<div></div>`
> 	- BTW the `_repr_svg_` style of encapsulating mime types is very nice and encourages the benefits of classing these elements in python code.

I think `write` was a mistake. That's something I aim to fix with the new API. (Thanks for confirming the choice of `_repr_svg_` , we think there may be better ways with things like more explicit mime type handlers but it's good to know it's useful nevertheless)

> 
> 3. display(*values, target="idname") Unfortunately display can't take an Element that has already been found like this:
>       ```python
>          block_div = Element("blocks")
>          display(*values, target=block_div)
>       ```
> 	 therefore more specific selection methods can't be used.

Ah, right! This is should be considered a bug in `display` actually... TY!

> 
> 4. BTW this works, which is great: `areas = block_div.element.querySelectorAll('[class*="area"]')`
> 	It returns a list of e.g.`[object SVGPathElement]` and only looks under the block_div for the classes with 'area' in the classname.
> 	Clearly working for SVG as well as HTML elements.
> 	This means the GPS style of doing CSS (IMHO arguably more programming oriented than the BEM model) is easy to do.

👍  I think this should also be improved in the new API since you'll basically Python objects that directly map to JS objects and you should almost never need to actually have to reference the internal `.elements` JS proxy object.

For instance, with the API I'm working on right now the code above should look something like:

`areas = block_div.query_all('[class*="area"]')` and should return Python objects that transparently map to JS objects.

> 5. Insertions and removals in the dynamic dom.
> I've been changing the UI after initial creation using the following successfully:
> ```python
> 	block_frag = '<div blahblah"> <div blah></div> </div>'
> 	harmony = Element("harmony")
> 	# for HTML
> 	color_blocks = harmony.element.querySelectorAll(".color-block")
> 	# - returns list of [object HTMLDivElement]
> 	# to remove
> 	color_blocks[2].remove()
> 	# and to add:
> 	color_blocks[i].insertAdjacentHTML("afterend", block_frag)
>     
> 	# for SVG
> 	svg = harmony.element.querySelector("#%s-ball-svg"%(id))
> 	# - returns an [object SVGSVGElement]
> 	# to remove
> 	svg.removeChild(svg.childNodes[-1]) # to pop off last svg element
> 	# to add is more complex and could do with a helper maybe
> 	newcircle = js.document.createElementNS('http://www.w3.org/2000/svg', 'circle');
> 	hue_circle_attr = [("cx","74"),("cy","38"),("r","6"),("transform","rotate(-90, 42, 42)"),("fill","none")]
> 	for attr,val in hue_circle_attr:
> 		newcircle.setAttribute(attr,val)
> 	svg.appendChild(newcircle)
> ```

Ok... with some flexibility, the code above should look more or less like this with the new interface. (consider that some things like `.children` below are pure speculation since I haven't thought of it or actually tried to implement it but throwing out ideas could be good for a dialogue and feedback)

```python
        from <whatevername_we_choose> import document, query
	block_frag = '<div blahblah"> <div blah></div> </div>'
	harmony = query(Element("harmony"))

	# for HTML
	color_blocks = harmony.query_all(".color-block")
	# - returns list of [Python Element Objects]

	# to remove
	color_blocks[2].remove()

	# and to add:
	color_blocks[i].insert("afterend", block_frag)
    
	# for SVG
	svg = harmony.query("#%s-ball-svg"%(id))
	# - returns an [object SVGSVGElement]

	# to remove
	svg.children.pop(-1)

	# to add is more complex and could do with a helper maybe
        newcircle = document.create(circle, namespace_uri='http://www.w3.org/2000/svg')
	hue_circle_attr = [("cx","74"),("cy","38"),("r","6"),("transform","rotate(-90, 42, 42)"),("fill","none")]
	for attr,val in hue_circle_attr:
		newcircle.set(attr, val)
	svg.append(newcircle)
```

Lmk if you have any feedback! Again, for the most part, this is still to be explored.

> 
> 6. py-events and @when
>  - `@when()` seems to be doing a similar job to js.addEventHandler() but has two issues for me:
>    - it fires on any child of the selector which means extra code is required to differentiate between parent and child with different classes, and to then act on the correct one.

Feels like a bug. Can you provide an example that we can use to reproduce?

>    - it works only on page load. So additions to the dom do not get event messages. This is a problem if using this mechanism to handle user events and I add another element to my array.
>    - This means I can't use @when on any element I intend to change in the dom.

I get what you mean (I think) but examples here would help too

>    
>  - py-events work well but, once inserted, are hard to change the parameters of. E.g.
>    - I can have a `py-click=foo(var1)` where var is calculated on intial insertion of the innerHtml.
>    - But its very hard to do any calculations when updating the argument obtained by target.getAttribute('py-click') because there is no notion of `this` when referring to the specific element I am updating.
>      - This means I have to add `Element()` selection and a complex selector based on my knowledge of the exact location of this element(was class only) in the dom, or I have to add an id to every element I wish to manipulate this way.
>      - Its also generating garbage to create and discard the `Element()` everytime this interaction occurs. 
>    - so to get around this for now, I have to add an id to all such elements and use a `global Element()` to avoid excessive garbage creation.
>    - so possibly a different mechanism is required, or its an argument for a global `this` to deal with events.

Same comment... would be good to have some examples. There are so many scenarios and talking while looking at code makes it really easier to discuss.

> 
>  - another possible garbage issue: Pyiodide has a `pyodide.ffi.wrappers.add_event_listener(elt, event, listener)` wrapper for JavaScript’s addEventListener() which automatically manages the lifetime of a JsProxy corresponding to the listener parameter.
>    - but what happens if I use micropython in the near future. Will I be leaving a lot of garbage events still attached even though the dom elements may have been removed ?
>    - IWBNI micropython based UI manipulation worked as well as Pyiodide.
> 
> I am not certain of what the 'right thing' to do is for any of the above.


Again, this is great feedback and food for thought. TY

**Reply by Neon22** [reply]
Thanks. Fantastic to see this getting a revisit.
I will add more very soon but just a quick note as I've been looking at scmoll etc and trying to work out a good way to add reactivity.

I think this will work: add a py-ref="foo" attribute to any html element we want to be reactive. When encountered in ```pyScriptMO ``` add the id(autogenerated if not defined) to a ```Reactive()``` class which is basically a dictionary keyed by the "foo" id. On the python side (which probably generated the html definition anyway) we can associate actions to do when the various listeners are triggered. To enable this the py-click (say) event hander should vector through this Reactive() class if py-ref was set.

I'll work on some more detailed examples but the bottom section of the main.py file has my workings on @when and py-event in:
- https://pyscript.com/@d5127013-0d63-43cd-9926-95380d21cf67/color-harmony-helper/latest
- but its my workings and is not setout to make it easy for you to see the experiments clearly.

**Comment by Neon22** [comment]
For @when examples. I submitted a documentation pull request https://github.com/pyscript/pyscript/pull/1563 which needed more work. Jeff gave good feedback to help me understand it better but it does indicate the problem with running only the first time the code is read and so you can't @when a new array (say) element.

**Reply by Neon22** [reply]
actually that doesn't ref the issue of @when executing only once because its a decorator. That was in a discord python-help thread. Here's the relevant snippet:
[17:11]Neon22: I am using the @when mechanism to respond to user input. when I create a new dom element dynamically it does not have the listener attached. Understandable. How do I attach it ? Should
 I have all the event code in an encapsulating class and then make a new instance, or put functions in functions, or ? Any hints as to the 'right thing' to do?
[00:23]Jeff Glass: As you say, @when only attaches listeners once, at the time the decorator executes. If you're creating DOM elements dynamically, perhaps instantiating the with py-[event] attributes is the way to go?

**Reply by Neon22** [reply]
The @when decorator not working example can be seen in the following:
 - https://pyscript.com/@d5127013-0d63-43cd-9926-95380d21cf67/color-harmony-helper/latest
 - clicking on a color block highlights the swatches that contain that color
 - If you add more blocks by incrementing the color count then clicking on a newly created color swatch does not fire the event and so no swatches are highlighted.

The reason being that the decoratored function is defined once (after I have created the initial inner_html).
```python
@when("click", selector=".color-block")
def highlight_swatches(evt):
   blahblahblah
```
 because this is not defined again, subsequent mods to the inner html (adding more color-blocks) means the @when action is not attached.

Perhaps the MutationObserver IIRC could be utilized in the decorator to fix that ?

**Reply by fpliger** [reply]
Ok... I think I understand the issue. It's quite useful for the new API in development...

I'll be working more on it this week and will follow up here as I explore it and digest the feedback. Thank you @Neon22 !

**Comment by Neon22** [comment]
wrt your first question on Needs: Initial page setup. Prebuilding html to populate the page:

Defining a new page of UI from an empty DOM or at least an empty div.
Three most common approaches(IMHO) might be:
1. python code to construct html srings and insert as inner_html
   E.g.
	```python
	blocks_div = Element("blocks")
    block_frag = '<div class="color-block" style="background-color: rgb(21, 52, 125);"><div class="color-code">#15347d</div></div>'
    inner_html = ""
    hue_img = '<img alt="Hueball" width="84px" height="84px" src="data:%s;base64,%s"/>' % (hueball['image']['mime'], hueball['image']['data'])
    hue_circle = '<circle cx="74" cy="38" r="6" transform="rotate(-90, 42, 42)"/>'
    for n in harmonies.keys():
        id = n.replace(" ","").lower()
        hue_circles = hue_circle*count
        ball_svg = '<svg  id="%s-ball-svg" width="84" height="84" style="stroke:#dddddd; fill:none;"><rect width="84" height="84"  fill="none"/>%s</svg>' % (id, hue_circles)
        spread_num = '<input id="%s-scrubbable" class="scrubbable-num" pattern="[0-9]+"  py-mouseup="generate_scheme(\'%s\', None,None,False,True)" py-input="generate_color_schemes()">' % (id, n)
        title_div = ' <div class="hm-title"><label for=%s>%s</label><div class="img-overlay-wrap"> %s %s %s</div></div> ' % (id, n, spread_num if id in spreadables else "", hue_img, ball_svg)
        #
        model_div = '<div id=%s class="color-harmony">' % (id) + title_div + '<div class="palette">'
        inner_html += model_div
        for tone in tone_labels:
            inner_html += '<div class="palette-row"><div class="label">%s</div>' %(tone)
            for j in range(count):
                inner_html += block_frag
            inner_html += "</div>" 
        inner_html += "</div></div>"
    blocks_div.element.innerHTML = inner_html
	```
	
	

2. Python classes which produce html and associated code - simulating a cohesive UI and keeping all the work together.
   E.g.
   - a class that deals with a popup menu
     - takes a list and a prefix name.
     - uses that prefix to construct an id to get the value from, to
	 - builds a select based (or other mechanism) popup menu of items with associated code for visibility and transitions.
	 - Thereby enabling a popup menu for single or multi-selection which is entirely parameterised and is constructed entirely in code.
     - Each instance of which creates a new section of popup html and functions for its behaviour.
  
3. A templating system like that in Schmol but using jinjaX.
  - Schmol code can be seen here for Pyscript: https://gitlab.levitnet.be/levit/schmol/-/blob/main/schmol/components.py
  - a one page on Jinjax - https://jinjax.scaletti.dev/
  - The purpose of which is (similar to above) create a single file component which can be used in multiple pages/positions without rewriting.

**Reply by Neon22** [reply]
OK just a note about templating. There must be something I have misunderstood because the author of jinjax thinks it would all have to be precompiled on a server and sent as a kind of SPA. (https://github.com/jpsca/jinjax/discussions/16#discussioncomment-6706464). His contention being its not really a suitable mechanism.

Of course Vue3 is running from a node.js server and so that's how their templating probably works also. But I naively thought it would be possible to insert html into the dom directly from a template referenced in python code and that template would be imported like a regular python module.

**Reply by Neon22** [reply]
found this templating engine written in python used by web2py. small - looks like what I was thinking
- https://github.com/web2py/yatl/tree/master/yatl

**Reply by Neon22** [reply]
OK here's a bit of fun.
 - imported yatl with idea of testing templating.
 - Works pretty well but `py-click=()` is not getting picked up when we mod the dom.
 - (Also have trouble with quoting but that's another problem)

An example (pretty minimal) here:
 - https://pyscript.com/@d5127013-0d63-43cd-9926-95380d21cf67/template-playground/latest
 - Just populating a dropdown using a javascript method.
 - You can see the dom getting modified but no py-click being registered.
 - Can compare with the Inspector looking at the similar dropdown here:
    -  https://pyscript.com/@d5127013-0d63-43cd-9926-95380d21cf67/color-harmony-helper/latest

**Comment by Neon22** [comment]
WRT your new API.
Its very good. I have been using a cloned copy of yours. It looks like "the right thing" to me.

**Reply by fpliger** [reply]
Awesome! Thanks for the feedback. I've been changing some concepts a bit and will be sharing the ideas around very soon. Hopefully it'll make it even more Pythonic and intuitive... :)

**Reply by Neon22** [reply]
Looking forward to it. I am happy to rewrite everything.

One last note about the py-click style of event.
If I have a structure like this:
```html
<div class="palette">
  <div class="palette-row">
    <div class="label">Classic</div>
    <div class="color-block" py-click="highlight_swatches(\'#AABBCC\')"
             style="background-color: rgb(1, 2, 3);" id="selected-color">
        <div class="color-code">#AABBCC</div></div>
    <div class="color-block" py-click="highlight_swatches(\'#ddeeff\')"
             style="background-color: rgb(4,5,6);">
        <div class="color-code">#ddeeff</div></div>
  </div>
</div>
```
Intention:
 - when user clicks on the color-block or color-code element:
    - the function `highlight_swatches()` is called with the color taken from the color-code element(always child of color-block).
    - the color code is only displayed if hovering. so either could be clicked by the user.

Problem:
 - As defined this is fine. BUt if the user performs some prior action which updates the values, there is no way to determine the new(current) value in that element.
    - by 'no way' I mean no simple way.
    - e.g. a this.accessor of some kind. Like highlight_swatches(this.child.value)

Ideally:
 - when updating an element - respecify the py-click
 OR
 - when updating an element - be able to edit the argument of the function `highlight_swatches()`
 AND/OR
 - have a this kind of accessor so we can dynamically determine the value in the pyclick definition

**Comment by Phelsong** [comment]
I was going to write a related proposal, but I'll throw it here.
Below is a working mockup of what I'm using for doing composable elements. There is probably a much better way of doing this, but this is what Ive worked out so far. 

## Button.py
```
from pyscript import Element

 class Button:

    _class_list: set = {"button", "uk-button"}
    _sub_components: list = []
    _inner_html: str = ""
    _inner_text: str

    def __init__(
        self,
        parent: Element,
        html_id: str,
        *,
        class_list: set = {},
        button_text: str = "I'm a button",
    ):
        """
        Parameters
        ----------
        html_id
        class_list
        """
        self.html_id = html_id
        self._class_list.update(class_list)
        cl = " ".join(x for x in self._class_list)
        self._inner_text = button_text
        self._inner_html = f'<button id={self.html_id} class="{cl}">{self._inner_text}</button>'

        self.element = self._Element(parent)

    # -------------------------------------------------------------------------

    def _repr_html_(self):
        return f"{self._inner_html}"

    # -------------------------------------------------------------------------

    def _Element(self, parent) -> Element:
        parent.write(self)
        return Element(self.html_id)
```

## Dashboard.py
snippet
```    
async def main() -> None:
    # ============
    FAKE_BODY: Element = Element("main")

    card_container: Element = Div(
        FAKE_BODY, "card-container", class_list={"card-container"}
    ).element
    card: Element = Card(card_container, "card-1").element
    svg: Element = SVG(card, "svg-1").element
    form: Element = Form(card, "input-form").element
    form_title: Element = Label(form, "input-form-title").element
    form_input: Element = Input(form, "input-form-input").element
    form_input_2: Element = Input(form, "input-form-input-2").element
    submit_button: Element = Button(parent=card, html_id="submit-button").element

    @when("click", selector="#submit-button")
    async def submit(event):
        print("submit")


ensure_future(main())
```
## Raw HTML
```
<!DOCTYPE pyscript>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <!-- ================================================================================== -->
    <link rel="manifest" href="/static/manifest.json" />
    <script defer src="/static/js/py_worker.js"></script>
    <link defer rel="stylesheet" href="/static/public/dashboard.css" />
    <script defer src="/static/pyscript/pyscript.js"></script>
    <!-- ================================================================================== -->
    <title>WIP</title>
  </head>
  <body>
    <!-- ==================================================== -->
    <py-terminal id="py-out"></py-terminal>
    <div id="main"></div>
    <!-- ==================================================== -->
    <py-config src="/web/py-script.toml"> </py-config>
    <!-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ -->
    <py-script src="/web/dashboard.py"> </py-script>
    <!-- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ -->
    <!-- ==================================================== -->
  </body>
</html>
```

## HTML output
![image](https://github.com/pyscript/pyscript/assets/98562418/c149bcff-0f1f-427b-8b8e-c82c8f00b1eb)
![image](https://github.com/pyscript/pyscript/assets/98562418/0c5fa472-a8a8-4298-97cf-10d8874b9804)



I built set of these templates built for about 15 different html types. Mostly because I wanted to work in a similar way to Kivy.

But I think ideally is having an Element.create method would be better.
`foo: Element = Element.create("div", id="foo", parent="body", **kwargs)`
This allows for easier per element manipulation, allows for much easier access to re-render individual component similar to HTMX, and most importantly pushes more of the written code into Python, instead of JS wrapper functions.

**Reply by Phelsong** [reply]
or alternately
```
foo: div = Element.div(parent="body", id="foo", **kwargs)
bar: button = Element.button(parent="foo", id="button-1", **kwargs) 
```
might perform better with AI tools (as my current personal implementation screams thru element building, since the context is very obvious) and having inline type hints would make it very clear what youre looking at.

**Comment by fpliger** [comment]
@Phelsong thanks for sharing your thoughts!

Actually, [here's an early version](https://fpliger.pyscriptapps.com/pyweb/latest/pydom.html) of the API I'm working on that I mentioned to @Neon22 above.

**Reply by Phelsong** [reply]
`new_div = pydom.create("div")`

So this is returning an Element or some other wrapper?

**Reply by Neon22** [reply]
@fpliger is there a link we can use to play with this atm ?
Or will it be in next release in a couple of days ?
I have stuff on pyscript.com if its accessible from there...

**Reply by fpliger** [reply]
@Phelsong it is an instance of the `pydom.Element` class. It's a Python class has a reference to the underlying HTML `Element` in the DOM. It is not a JS `Element`

@Neon22 yes! and Yes (at least I'm trying to get it included in the release in time...). [Here's a quick example you can use ](https://__fabio__.pyscriptapps.com/purple-sound/latest/)to experiment with it. You can clone the project and write your own code or just copy the `fetch` setting from the config TOML file. You may see CORS issues in the deployed app (I think it's a known bug that needs fixing soon) but if you clone it, it should work fine within the iFrame in pyscript.com

**Reply by Phelsong** [reply]
I like that a lot then!

**Comment by Neon22** [comment]
@fpliger Is it possible to modify your `BaseElement create()` method so that if we handed it an `Element` it would get the arguments it needs from that ?
What I'm getting at it is:
 - if we wrote a new class that was derived from BaseElement (or used some other mechanism you thought would work better)
 - that made a (say) button (in the spirit of what @Phelsong posted above) with children/attributes/etc and behaviours like `py-click` etc defined
 - then we could build up a sophisticated ui by defining components separately.
 - then we could build a reactive etc approach on top of that.

There might be a better way of course. Ideas @Phelsong ??

**Reply by fpliger** [reply]
> @fpliger Is it possible to modify your BaseElement create() method so that if we handed it an Element it would get the arguments it needs from that ?

@Neon22 , you mean a JS element ?

**Reply by Neon22** [reply]
Imagine I wrote a button component for me to use ( below I have used @Phelsong idea but in your new pydom package).
```python
class MyButton(object):
    def __init__(self, parent=None, html_id=None, classes=None,
                label="Foo"):
        self.html_id = html_id
        self.parent = parent if parent else pydom
        self.classes = classes
        self.label = label

    def create(self):
        self.el = self.parent.create("button", classes=self.classes, html=self.label)
        self.el.id = self.html_id
        return self

    def __repr__(self):
        return "<MyButton '%s' id=%s classes='%s', parent=%s" %(self.label, self.html_id, " ".join(x for x in self.classes), self.parent )
    def _repr_html_(self):
        return self.el._element

a = MyButton(parent=new_div, label="One", classes=["bar","baz"], html_id="foo").create()
print(a)
```

It might be this is the best way.
Or it might be that I should subclass your `class Element(BaseElement)` and build my own component there.
What is your opinion ?

**Reply by Neon22** [reply]
I should have added pyclick-event and @when as well to show what I mean but hope you get idea...
by subclass I mean instead of `class MyButton(object):` it would be `class MyButton(pydom.Element):`

**Reply by Phelsong** [reply]
yeah that's what I got from reading it. Since the pydom is a direct replacement for the Element API. Syntactically seems fine, I can't think of anything that could easily be done to consolidate it without excessive abstraction. 

I think there will be a lot more work that needs to be done (over time) to get all the functionality.

Overall I feel like it accomplishes what I would hope for, tho I do have some thoughts as far as extending the functionality I'm mulling thru.
